consensus/src/runway/collection.rs

```diff
@@ -0,0 +1,536 @@
+    for Error<H, D, S>
+{
+    fn from(_: SignatureError<NewestUnitResponse<H, D, S>, S>) -> Self {
+        Error::WrongSignature
+    }
+}
+
+/// The status of an ongoing collection.
+#[derive(PartialEq, Debug)]
+pub enum Status {
+    /// Received less than threshold responses.
+    Pending,
+    /// Received at least threshold responses.
+    Ready(Round),
+    /// Received all possible responses.
+    Finished(Round),
+}
+
+/// Initial unit collection to figure out at which round we should start unit production.
+/// Unfortunately this isn't quite BFT, but it's good enough in many situations.
+pub struct Collection<'a, MK: KeyBox> {
+    keychain: &'a MK,
+    validator: &'a Validator<'a, MK>,
+    starting_round: Round,
+    responders: HashSet<NodeIndex>,
+    threshold: NodeCount,
+    salt: Salt,
+}
+
+impl<'a, MK: KeyBox> Collection<'a, MK> {
+    /// Create a new collection instance ready to collect responses.
+    /// The returned salt should be used to initiate newest unit requests.
+    pub fn new(
+        keychain: &'a MK,
+        validator: &'a Validator<'a, MK>,
+        threshold: NodeCount,
+    ) -> (Self, Salt) {
+        let salt = generate_salt();
+        let responders = [keychain.index()].into();
+        (
+            Collection {
+                keychain,
+                validator,
+                starting_round: 0,
+                responders,
+                threshold,
+                salt,
+            },
+            salt,
+        )
+    }
+
+    /// Process a response to a newest unit request.
+    pub fn on_newest_response<H: Hasher, D: Data>(
+        &mut self,
+        unchecked_response: UncheckedSigned<NewestUnitResponse<H, D, MK::Signature>, MK::Signature>,
+    ) -> Result<Status, Error<H, D, MK::Signature>> {
+        let response = unchecked_response.check(self.keychain)?.into_signable();
+        if response.salt != self.salt {
+            return Err(Error::SaltMismatch(self.salt, response.salt));
+        }
+        if let Some(unchecked_unit) = response.unit {
+            let checked_signed_unit = self.validator.validate_unit(unchecked_unit)?;
+            let checked_unit = checked_signed_unit.as_signable();
+            if checked_unit.creator() != self.keychain.index() {
+                return Err(Error::ForeignUnit(checked_unit.creator()));
+            }
+            let starting_round_candidate = checked_unit.round() + 1;
+            if starting_round_candidate > self.starting_round {
+                self.starting_round = starting_round_candidate;
+            }
+        }
+        self.responders.insert(response.responder);
+        Ok(self.status())
+    }
+
+    /// The salt associated with this collection instance.
+    pub fn salt(&self) -> Salt {
+        self.salt
+    }
+
+    /// The current status of the collection.
+    pub fn status(&self) -> Status {
+        use Status::*;
+        let responders = NodeCount(self.responders.len());
```

Comment:
Small nit: In two ifs here you compare in different way (one by creating `NodeCount`, second by taking `.0`). Maybe here could be something like (`NodeCount` has `PartialOrd` derived) 
```
let responder_count = NodeCount(self.responders.len());
if responder_count == self.keychain.node_count() {
    return Finished(self.starting_round);
}
if responder_count >= self.threshold {
    return Ready(self.starting_round);
}
```

Reply:
Yup, you are completely right, no idea why I thought I had to touch the `.0`s. Fixed.